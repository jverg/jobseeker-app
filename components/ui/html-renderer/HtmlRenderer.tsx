import React from 'react';
import sanitizeHtml from 'sanitize-html';
import styles from './HtmlRenderer.module.less';

type HtmlRendererType = {
  html: string;
  className?: string;
  testId?: string;
};

const HtmlRenderer: React.FC<HtmlRendererType> = ({ html, className = '', testId }) => {
  /**
   * In order to make the iframes (videos) responsive
   * we have to add a class on the parent (p tag) element
   * to target it through css
   */
  const addClassToIframeParent = html.replace(/<p><iframe/g, '<p class="iframe-container"><iframe loading="lazy"');

  /**
   * Html sanitization
   * All default tags are allowed, adding
   * - <iframe> for video embeds
   * - <img> for images
   * - <a> for links
   * The allowed hostnames for iframes are youtube and vimeo
   */
  const sanitizedHtml = sanitizeHtml(addClassToIframeParent, {
    allowedTags: sanitizeHtml.defaults.allowedTags.concat(['iframe', 'img']),
    allowedAttributes: {
      iframe: ['src', 'width', 'height', 'allowfullscreen', 'loading'],
      p: ['class'],
      img: ['src', 'alt', 'width', 'height'],
      a: ['href', 'target'],
    },
    allowedIframeHostnames: ['www.youtube.com', 'player.vimeo.com'],
  });

  return (
    <div
      className={`${styles.renderer} ${className}`}
      dangerouslySetInnerHTML={{ __html: sanitizedHtml }}
      data-testid={`html-renderer-${testId}`}
    />
  );
};

HtmlRenderer.defaultProps = {
  className: undefined,
  testId: undefined,
};

export default HtmlRenderer;
