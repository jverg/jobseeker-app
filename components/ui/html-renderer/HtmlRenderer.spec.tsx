import React from 'react';
import HtmlRenderer from '@components/ui/html-renderer/HtmlRenderer';
import { cleanup, render, screen } from '@testing-library/react';

describe('HtmlRenderer', () => {
  afterEach(() => {
    cleanup();
  });
  it('should render a basic paragraph given as html', () => {
    render(<HtmlRenderer html="<p>Test paragraph</p>" testId="test" />);
    expect(screen.getByTestId('html-renderer-test')).toBeVisible();
    expect(screen.getByText('Test paragraph')).toBeVisible();
    expect(screen.getByTestId('html-renderer-test')).toHaveTextContent('Test paragraph');
  });
  it('should render a more complex html and escape dangerous tags', () => {
    const complexHtml = `
      <h2>This is the heading h2</h2>
      <h3>This is heading h3</h3>
      <p>Bla bla bla <a href="https://www.google.com" target="_blank" rel="noopener">google</a> bla bla bla <strong>blablabla </strong>bla <em>blabla </em><span style="text-decoration: underline;">bla</span></p>
      <ul>
      <li>1</li>
      <li>2</li>
      <li>3</li>
      </ul>
      <p>Bla bla bla</p>
      <p><img src="https://upload.wikimedia.org/wikipedia/commons/thumb/b/b6/Imâ€¦obile_phone.png/1200px-Image_created_with_a_mobile_phone.png" alt="" width="257" height="193" /> <img src="https://employerfiles.blob.core.windows.net/useruploads/1005/branded-company/images/2d79deeba1e03.jpg" alt="" width="290" height="192" /></p>
      <p><iframe style="width: 547px; height: 222px;" title="YouTube video player" src="https://www.youtube.com/embed/womsDCDcEEM" width="547" height="222" frameborder="0" allowfullscreen="allowfullscreen"></iframe></p>
      <script>alert('hello')</script>
    `;

    const { container } = render(<HtmlRenderer html={complexHtml} testId="test" />);
    expect(container).toMatchSnapshot();
  });
});
