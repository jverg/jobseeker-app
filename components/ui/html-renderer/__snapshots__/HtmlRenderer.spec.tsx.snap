// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HtmlRenderer should render a more complex html and escape dangerous tags 1`] = `
<div>
  <div
    class="undefined "
    data-testid="html-renderer-test"
  >
    
      
    <h2>
      This is the heading h2
    </h2>
    
      
    <h3>
      This is heading h3
    </h3>
    
      
    <p>
      Bla bla bla 
      <a
        href="https://www.google.com"
        target="_blank"
      >
        google
      </a>
       bla bla bla 
      <strong>
        blablabla 
      </strong>
      bla 
      <em>
        blabla 
      </em>
      <span>
        bla
      </span>
    </p>
    
      
    <ul>
      
      
      <li>
        1
      </li>
      
      
      <li>
        2
      </li>
      
      
      <li>
        3
      </li>
      
      
    </ul>
    
      
    <p>
      Bla bla bla
    </p>
    
      
    <p>
      <img
        alt=""
        height="193"
        src="https://upload.wikimedia.org/wikipedia/commons/thumb/b/b6/Imâ€¦obile_phone.png/1200px-Image_created_with_a_mobile_phone.png"
        width="257"
      />
       
      <img
        alt=""
        height="192"
        src="https://employerfiles.blob.core.windows.net/useruploads/1005/branded-company/images/2d79deeba1e03.jpg"
        width="290"
      />
    </p>
    
      
    <p
      class="iframe-container"
    >
      <iframe
        allowfullscreen="allowfullscreen"
        height="222"
        loading="lazy"
        src="https://www.youtube.com/embed/womsDCDcEEM"
        width="547"
      />
    </p>
    
      
    
  </div>
</div>
`;
